common/src/main/java/dev/ftb/mods/ftbchunks/client/mapicon/EntityImageIcon.java

```diff
@@ -0,0 +1,113 @@
+package dev.ftb.mods.ftbchunks.client.mapicon;
+
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import dev.ftb.mods.ftblibrary.icon.Icon;
+import dev.ftb.mods.ftblibrary.icon.ImageIcon;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.renderer.texture.SimpleTexture;
+import net.minecraft.resources.ResourceLocation;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Optional;
+
+public class EntityImageIcon extends Icon {
+    @Nullable
+    private final Slice mainSlice;
+    private final List<ChildIconData> children;
+    private final Icon mainIcon;
+    private final List<Icon> childIcons;
+
+    public EntityImageIcon(ResourceLocation mainTexture, @Nullable Slice mainSlice, List<ChildIconData> children) {
+        this.mainSlice = mainSlice;
+        this.children = children;
+
+        mainIcon = createIcon(mainTexture, mainSlice);
+        childIcons = children.stream().map(childIconData -> createIcon(childIconData.texture.orElse(mainTexture), childIconData.slice)).toList();
+    }
+
+    @Override
+    public void draw(GuiGraphics graphics, int x, int y, int width, int height) {
+        PoseStack pose = graphics.pose();
+        pose.pushPose();
+        pose.translate(x, y, 0);
+
+        float drawWidth = mainSlice == null ? width : mainSlice.width;
+        float drawHeight = mainSlice == null ? height : mainSlice.height;
+
+        float scaleX = width / drawWidth;
+        float scaleY = height / drawHeight;
+
+        pose.scale(scaleX, scaleY, 1);
+
+        mainIcon.draw(graphics, 0, 0, (int) drawWidth, (int) drawHeight);
+
+        for (int i = 0; i < children.size(); i++) {
+            ChildIconData child = children.get(i);
+            Icon icon = childIcons.get(i);
+
+            pose.pushPose();
+            child.offset.ifPresent(offset -> pose.translate(offset.x, offset.y, 0));
+            icon.draw(graphics, 0, 0, child.slice.width, child.slice.height);
+            pose.popPose();
+        }
+
+        pose.popPose();
+    }
+
+
+    public record Offset(int x, int y) {
+
+        public static final Codec<Offset> CODEC = RecordCodecBuilder.create(builder -> builder.group(
+                Codec.INT.fieldOf("x").forGetter(o -> o.x),
+                Codec.INT.fieldOf("y").forGetter(o -> o.y)
+        ).apply(builder, Offset::new));
+
+    }
+
+    public record Slice(
+            int x,
+            int y,
+            int width,
+            int height) {
+
+        public static final Codec<Slice> CODEC = RecordCodecBuilder.create(builder -> builder.group(
```

Comment:
Can't you combine codecs to create a combination of the offset and width and height ones?